### DISCUSSION TITLE
cisco asav example

### DISCUSSION THREAD

**Discussion by lax-pati** [ipspace/netlab#1600] [General]
Hello,

Thanks for coming up with such an amazing project..no words to express my gratitude as how easy you guys have made the life of many network explorers :) 

I am not familiar with FW config and wondering if there is any sample lab or an example to use cisco asav with netlab topology. i checked the examples but could not find any info on that. I have created a asa vagrant box and planning to use it to setup with few routers and hosts along with asav. Any info would help greatly.... Thank you

**Comment by jbemmel** [comment]
You can do:
```
netlab up tests/integration/initial/01-interfaces.yml -d asav
```

Build instructions for the appliance: https://netlab.tools/labs/asav/

**Reply by lax-pati** [reply]
Thanks, yes, my asav vagrant box is up.. but i am looking for any sample topology file example to stich asav with other routers. netlab has given beautiful examples without breaking head on getting the configs i.e ospf, bgp etc. wondering if something similar exists for using asav :)

**Comment by lax-pati** [comment]
any help please @jbemmel  @steffann  @ipspace  ?

**Comment by lax-pati** [comment]
Appreciate that. Few hosts sitting behind routers(ceos,iosv kind) attached
to assv, allowing hosts to communicate (internal networks). One host acting
like external network, if it tries to ping, I guess assv would block it.
Please correct me if that's not how asav supposed to work.

h1--r1-asav-r2-h2 ping should work. h3-r3-asav-r1-h1 ping should not work.
Is this ok?

If this doesn't make sense, any topology that you think appropriate could
help, just that there are few routers and hosts traffic through asav..

Thanks for your willingness to help

**Comment by ipspace** [comment]
Here's a sample topology https://github.com/ipspace/netlab-examples/tree/master/multi-platform/asav, details described in https://blog.ipspace.net/2024/12/netlab-asav-topology/.

Routing works, but I don't know enough about ASAv to configure the firewall/security bits that would allow H1 to ping X.

**Reply by lax-pati** [reply]
Thats so sweet @ipspace , thanks a lot for your help :)

**Reply by lax-pati** [reply]
Hi @ipspace  tried using the example, but getting into ssh issues. used the options suggested in your blog i.e tried adding in topology file and also modifying the ssh config to accept old algos but no luck. this is what i added in .ssh/config :
Host *
    KexAlgorithms +diffie-hellman-group1-sha1,diffie-hellman-group-exchange-sha1
    Ciphers +aes128-cbc,3des-cbc
    PubkeyAcceptedAlgorithms +ssh-rsa
    HostkeyAlgorithms +ssh-rsa

vagrant ssh also did not work. its complaining that "/root/.ssh/config: terminating, 1 bad configuration options".  i have netlab version 1.9.2

virsh console of asav throwing
"ERROR: /etc/ssh/lina_sshd.conf: No such file or directory".  Any pointers on resolving this please ?

**Comment by ipspace** [comment]
@lax-pati: I always had problems with '+x' in .ssh/config. According to the documentation it should work, but (sometimes?) does not. That's why I went for **netlab_ssh_args**.

This might work:

* Set **netlab_ssh_args** in topology file
* Start the lab and do **netlab inspect defaults.devices.asav.group_vars** to verify the **netlab_ssh_args** variable was set as expected.
* Execute **netlab connect -v fw** to see the SSH command that will be executed
* Extract the SSH command from the weird printout (sorry about that), replace LogLevel=ERROR with LogLevel=INFO and execute it.
* You should get an error message telling you what algorithms ASAv wants to use. Use that to adjust the **netlab_ssh_args** variable.
* Report what worked ;)

**Reply by lax-pati** [reply]
@ipspace  Thank you... now a weird thing. netlab connect -v fw working when the ssh args is included in topology file, not sure what was missing yday !!  with that, did netlab create with which asav came up fine. Then did " netlab up --snapshot -vv --no-config", its complaining about the bridge now :)  

```
┌──────────────────────────────────────────────────────────────────────────────────┐
│ STARTING libvirt nodes                                                           │
└──────────────────────────────────────────────────────────────────────────────────┘
pre-start hook for libvirt
run_command executing: ['vagrant', 'status', '--machine-readable']
run_command executing: ['virsh', 'net-destroy', 'nl_mgmt_18']
run_command executing: ['virsh', 'net-undefine', 'nl_mgmt_18']
run_command executing: ['sudo', 'ip', 'link', 'delete', 'nl_mgmt_18']
[CREATED] creating libvirt management network nl_mgmt_18
run_command executing: ['virsh', 'net-define', '/tmp/tmprb9vwzo7']
provider libvirt: executing vagrant up --provider libvirt
run_command executing: vagrant up --provider libvirt
Bringing machine 'fw' up with 'libvirt' provider...
==> fw: Checking if box 'cisco/asav' is up to date...
==> fw: Creating image (snapshot of base box volume).
==> fw: Creating domain with the following settings...
==> fw:  -- Name:              ml_18__fw
==> fw:  -- Domain type:       kvm
==> fw:  -- Cpus:              1
==> fw:
==> fw:  -- Feature:           acpi
==> fw:  -- Feature:           apic
==> fw:  -- Feature:           pae
==> fw:  -- Memory:            2048M
==> fw:  -- Management MAC:    08:4f:a9:00:00:01
==> fw:  -- Loader:
==> fw:  -- Base box:          cisco/asav
==> fw:  -- Storage pool:      default
==> fw:  -- Image:             /var/lib/libvirt/images/ml_18__fw.img (9G)
==> fw:  -- Volume Cache:      default
==> fw:  -- Kernel:
==> fw:  -- Initrd:
==> fw:  -- Graphics Type:     vnc
==> fw:  -- Graphics Port:     -1
==> fw:  -- Graphics IP:       127.0.0.1
==> fw:  -- Graphics Password: Not defined
==> fw:  -- Video Type:        cirrus
==> fw:  -- Video VRAM:        9216
==> fw:  -- Sound Type:
==> fw:  -- Keymap:            en-us
==> fw:  -- TPM Path:
==> fw:  -- INPUT:             type=mouse, bus=ps2
==> fw: Creating shared folders metadata...
==> fw: Starting domain.
==> fw: Waiting for domain to get an IP address...
==> fw: Waiting for SSH to become available...
==> fw: Configuring and enabling network interfaces...
libvirt lab has started, fixing Linux bridges
run_command executing: ['virsh', 'net-info', 'nl_mgmt_18']
run_command executing: ['virsh', 'net-info', 'ml_18_2']
... network ml_18_2 maps into virbr3
run_command executing: ['sudo', 'sh', '-c', 'echo 0x4000 >/sys/class/net/virbr3/bridge/group_fwd_mask']
... setting LLDP enabled flag on virbr3
run_command executing: ['virsh', 'net-info', 'ml_18_3']
... network ml_18_3 maps into virbr4
run_command executing: ['sudo', 'sh', '-c', 'echo 0x4000 >/sys/class/net/virbr4/bridge/group_fwd_mask']
... network ml_18_3 maps into virbr4
run_command executing: ['sudo', 'sh', '-c', 'echo 0x4000 >/sys/class/net/virbr4/bridge/group_fwd_mask']
... setting LLDP enabled flag on virbr4

┌──────────────────────────────────────────────────────────────────────────────────┐
│ STARTING clab nodes                                                              │
└──────────────────────────────────────────────────────────────────────────────────┘
Recreating clab-augment.yml configuration file for clab provider
[CREATED] provider configuration file: clab-augment.yml
[MAPPED]  clab_files/h1/hosts to h1:/etc/hosts (from templates/provider/clab/linux/hosts.j2)
[MAPPED]  clab_files/h2/hosts to h2:/etc/hosts (from templates/provider/clab/linux/hosts.j2)
[MAPPED]  clab_files/x/hosts to x:/etc/hosts (from templates/provider/clab/linux/hosts.j2)
pre-start hook for Containerlab - create any bridges
run_command executing: ['brctl', 'show', 'ml_18_1']
Linux bridge ml_18_1 already exists, skipping
provider clab: executing sudo -E containerlab deploy --reconfigure -t clab-augment.yml
run_command executing: sudo -E containerlab deploy --reconfigure -t clab-augment.yml
INFO[0000] Containerlab v0.59.0 started
INFO[0000] Parsing & checking topology file: clab-augment.yml
INFO[0000] Removing /home/labs/play/clab-ml_18 directory...
INFO[0000] Creating docker network: Name="nl_mgmt_18", IPv4Subnet="192.168.18.0/24", IPv6Subnet="", MTU=0
Error: bridge "ml_18_1" referenced in topology but does not exist
bridge "ml_18_1" referenced in topology but does not exist
bridge "ml_18_1" referenced in topology but does not exist
Error executing sudo -E containerlab deploy --reconfigure -t clab-augment.yml:
  Command '['sudo', '-E', 'containerlab', 'deploy', '--reconfigure', '-t', 'clab-augment.yml']' returned non-zero exit status 1.
[FATAL]   netlab up: sudo -E containerlab deploy --reconfigure -t clab-augment.yml failed, aborting...

```
I am running this with multilab, tried changing to different subnet, same issue. used ovs instead of linux bridge ( checked in some thread that ovs can be used instead of linux bridge, so gave a try ! ) , getting similar issue. I have no idea about this bridging, so please help

**Reply by ipspace** [reply]
Can you please post the topology file? Are you using Ubuntu or something else (although I'm guessing it might be a netlab issue)? Can you try it out without multilab plugin?

Also, as weird as that sounds, there might be a non-zero chance that shutting down the lab and rebooting the server might help ;)

**Reply by ipspace** [reply]
One more thing -- if the lab fails to start after rebooting the server, please execute "brctl show ml_18_1" (or whatever brctl command is executed during the containerlab start phase) and post the results.

It looks like the results of the "brctl" command indicate the bridge is there, whereas in reality it's not.

**Reply by ipspace** [reply]
Finally, if you're on Ubuntu 18.x, you might be experiencing a variant of this: https://github.com/ipspace/netlab/discussions/1310

**Reply by lax-pati** [reply]
oh god, this looks exactly similar.. let me try tweaking the way we specify links as captured in other issue.

```- h1-h2-int
- int-fw
- fw:
  ext:
    bgp.default_originate: True
- ext:
    bgp.advertise: False
  x:
```

**Reply by ipspace** [reply]
> oh god, this looks exactly similar.. let me try tweaking the way we specify links as captured in other issue.

Let's figure out what's wrong with the current setup first (as in "don't change the parameters in the midst of an experiment"):

* What version of Linux are you using?
* What does "brctl show X" and "brctl show X && echo OK" produce for the "missing" bridge?

**Reply by lax-pati** [reply]
@ipspace  hmm ok, But i changed the topo to below, kept the linux bridge and put on a different multilab id, it worked!!! no ssh issues of any sort. did netlab initial which worked like a charm.... yes, my system is on ubuntu 18.04.. this behavior is so weird, a puzzle to me !!

```
links:
- h1-int
- h2-int
- int-fw
- fw:
  ext:
    bgp.default_originate: True
- ext:
    bgp.advertise: False
  x:
```

**Reply by ipspace** [reply]
Nice to see you got it up and running. Anyway, **netlab** needs to create an extra Linux bridge to implement multi-access networks for containerlab (which natively supports only P2P links), and to figure out if the bridge already exists, it executes the **brctl show _name_** command.

That command should return an error code if the bridge does not exist (and so netlab will go and create it), but sometimes (on Ubuntu 18.04) it generates an error but returns OK status. That's clearly a bug in **brctl** that has been fixed in the meantime, but it's been impossible for me to reproduce it as I stopped using Ubuntu 18.04 a long while ago (and supposedly it doesn't happen consistently).

**Reply by lax-pati** [reply]
Ok, thank you @ipspace for your quick responses :)  I am sure you have enough on your plate, but u took time to respond to my asks, much appreciated...  i will check what you posted previously when i tear this lab and go back to the original topo. But in other issue that you shared, things worked with ovs, but not for me even with ovs.. so seems its high time for upgrading ubuntu :)

**Reply by lax-pati** [reply]
Hi @ipspace  topology is up and working. You had mentioned about the config to get pings from h1 to x are not there. This is the piece to get it working. Pings from h1 to x would work but fw would block pings from x to h1..

configured "security-level 100" on internal network port i.e gi0/0  and  an acl to allow pings from internal to external. By default, asa allows traffic from higher security level to lower security level i.e from 100 -> 0, as external port security-level is 0, which is not changed.  Not sure if this is the ideal config, but it works !!

```
access-group ALLOW_INTERNAL in interface internal
access-list ALLOW_INTERNAL extended permit ip 172.16.0.0 255.255.0.0 any
```

**Reply by ipspace** [reply]
Thank you!

**Reply by ipspace** [reply]
Updated the blog post (https://blog.ipspace.net/2024/12/netlab-asav-topology/). Thanks again!
